### ISSUE TITLE
Issues in rendering TikZ patterns

### ISSUE THREAD

**Issue by parpalak** [mgieseki/dvisvgm#158]
Hello Martin,

I've encountered some problems with TikZ patterns in dvisvgm when debugging an issue: https://github.com/parpalak/i.upmath.me/issues/13

I took a LaTeX example from here: https://newbedev.com/overwriting-fill-patterns-in-tikz

```latex
\documentclass[tikz]{standalone}
\usetikzlibrary{patterns}
\begin{document}
\begin{tikzpicture}
\fill[yellow!20] (-1,-1) rectangle (5,5);
\draw[pattern=dots, pattern color=green] (0, 0) rectangle (4, 4) (2, 2) circle[radius=1];
\path[pattern=bricks,pattern color=brown] (2,2) circle (1);
\end{tikzpicture}
\end{document}
```

When converted with latex and dvisvgm (TEX -> DVI -> SVG) it is rendering without any figures with patterns:

```svg
<?xml version='1.0' encoding='UTF-8'?>
<!-- This file was generated by dvisvgm 2.11.1 -->
<svg version='1.1' xmlns='http://www.w3.org/2000/svg' xmlns:xlink='http://www.w3.org/1999/xlink' width='170.0824pt' height='170.0779pt' viewBox='-72 -71.9998 170.0824 170.0779'>
<defs>
<pattern id='pat0-ff00' x='-.99628' y='-.99628' width='2.98883' height='2.98883' viewBox='-.99628 -.99628 2.98883 2.98883' patternUnits='userSpaceOnUse' patternTransform='matrix(1 0 0 -1 -43.6536 69.7322)'>
<clipPath id='pc0'>
<rect x='-.99628' y='-.99628' width='1.99256' height='1.99256'/>
</clipPath>
<g clip-path='url(#pc0)'>
<path d='M.5 0C.5 .273438 .273438 .5 0 .5S-.5 .273438-.5 0S-.273438-.5 0-.5S.5-.273438 .5 0Z'/>
<path d='M-43.652345 69.730468V-43.6558H69.7344V69.730468ZM41.3867 13.0391C41.3867-2.6172 28.6953-15.3086 13.0391-15.3086C-2.6172-15.3086-15.3086-2.6172-15.3086 13.0391C-15.3086 28.6953-2.6172 41.3867 13.0391 41.3867C28.6953 41.3867 41.3867 28.6953 41.3867 13.0391Z' fill='url(#pat0-ff00)'/>
<path d='M-43.652345 69.730468V-43.6558H69.7344V69.730468ZM41.3867 13.0391C41.3867-2.6172 28.6953-15.3086 13.0391-15.3086C-2.6172-15.3086-15.3086-2.6172-15.3086 13.0391C-15.3086 28.6953-2.6172 41.3867 13.0391 41.3867C28.6953 41.3867 41.3867 28.6953 41.3867 13.0391Z' stroke='#000' fill='none' stroke-width='.3985' stroke-miterlimit='10'/>
<path d='M41.3867 13.0391C41.3867-2.6172 28.6953-15.3086 13.0391-15.3086C-2.6172-15.3086-15.3086-2.6172-15.3086 13.0391C-15.3086 28.6953-2.6172 41.3867 13.0391 41.3867C28.6953 41.3867 41.3867 28.6953 41.3867 13.0391Z' fill='url(#pat1-bf8040)'/>
</g>
</pattern>
<pattern id='pat1-bf8040' x='-.99628' y='-.99628' width='11.3387' height='11.3387' viewBox='-.99628 -.99628 11.3387 11.3387' patternUnits='userSpaceOnUse' patternTransform='matrix(1 0 0 -1 -43.6536 69.7322)' overflow='visible'>
<clipPath id='pc1'>
<rect x='-.99628' y='-.99628' width='15.16968' height='15.16968'/>
</clipPath>
<g clip-path='url(#pc1)'>
<path d='M0 2.83594H11.3398M0 8.50391H11.3398M2.83594 0V2.83594M8.50391 2.83594V8.50391M2.83594 8.50391V11.3398' stroke='#bf8040' fill='none' stroke-width='.79701' stroke-miterlimit='10'/>
</g>
</pattern>
</defs>
<g id='page1'>
<path d='M-72 98.0781V-71.9998H98.0824V98.0781Z' fill='#ffc'/>
</g>
</svg>
```

![image](https://user-images.githubusercontent.com/9402484/128758484-fe5fab38-4551-4665-b4fc-c14d79f2543c.png)

But when the same DVI is converted to PS and then to PDF (with dvips and ps2pdf) it looks right:
[src2.pdf](https://github.com/mgieseki/dvisvgm/files/6956748/src2.pdf)

![image](https://user-images.githubusercontent.com/9402484/128758664-edf70b4e-65e9-4fe1-ae54-2da6c27d90df.png)

The issue can be reproduced in TeX Live 2021.

Could you please check why the convertation to SVG fails? Is it possible to use patterns in TikZ?

**Comment by mgieseki** [comment]
Hi Roman,

yes, you're right, that's a known issue with version 2.11.1. I've already fixed it a couple of month ago (e1b1e71f286526582e9cc6f0e71232f48995063a) but haven't found enough time to publish a new release yet. If possible, you could build a fixed version from the latest sources.

There seems to be another issue, however. While the shapes convert correctly, the green color of the dotted pattern is missing. I'll have a look at this in the next couple of days.

```svg
<?xml version='1.0' encoding='UTF-8'?>
<!-- This file was generated by dvisvgm 2.11.1 -->
<svg version='1.1' xmlns='http://www.w3.org/2000/svg' xmlns:xlink='http://www.w3.org/1999/xlink' width='170.08pt' height='170.08pt' viewBox='-72 -72 170.08 170.08'>
<defs>
<pattern id='pat0-ff00' x='-1' y='-1' width='2.99' height='2.99' viewBox='-1 -1 2.99 2.99' patternUnits='userSpaceOnUse' patternTransform='matrix(1 0 0 -1 -43.65 69.73)'>
<clipPath id='pc0'>
<rect x='-1' y='-1' width='1.99' height='1.99'/>
</clipPath>
<g clip-path='url(#pc0)'>
<path d='m.5 0c0 .27-.23 .5-.5 .5s-.5-.23-.5-.5s.23-.5 .5-.5s.5 .23 .5 .5z'/>
</g>
</pattern>
<pattern id='pat1-bf8040' x='-1' y='-1' width='11.34' height='11.34' viewBox='-1 -1 11.34 11.34' patternUnits='userSpaceOnUse' patternTransform='matrix(1 0 0 -1 -43.65 69.73)' overflow='visible'>
<clipPath id='pc1'>
<rect x='-1' y='-1' width='15.17' height='15.17'/>
</clipPath>
<g clip-path='url(#pc1)'>
<path d='m0 2.84h11.34m-11.34 5.67h11.34m-8.5-8.5v2.84m5.67 0v5.67m-5.67 0v2.84' stroke='#bf8040' fill='none' stroke-width='.8' stroke-miterlimit='10'/>
</g>
</pattern>
</defs>
<g id='page1'>
<path d='m-72 98.08v-170.08h170.08v170.08z' fill='#ffc'/>
<path d='m-43.65 69.73v-113.39h113.39v113.39zm85.04-56.69c0-15.66-12.69-28.35-28.35-28.35s-28.35 12.69-28.35 28.35s12.69 28.35 28.35 28.35s28.35-12.69 28.35-28.35z' fill='url(#pat0-ff00)'/>
<path d='m-43.65 69.73v-113.39h113.39v113.39zm85.04-56.69c0-15.66-12.69-28.35-28.35-28.35s-28.35 12.69-28.35 28.35s12.69 28.35 28.35 28.35s28.35-12.69 28.35-28.35z' stroke='#000' fill='none' stroke-width='.4' stroke-miterlimit='10'/>
<path d='m41.39 13.04c0-15.66-12.69-28.35-28.35-28.35s-28.35 12.69-28.35 28.35s12.69 28.35 28.35 28.35s28.35-12.69 28.35-28.35z' fill='url(#pat1-bf8040)'/>
</g>
</svg>
```

**Comment by parpalak** [comment]
Thanks for the quick answer. I'd prefer to wait a fresh release of TeX Live. When I tried to build from source last time (https://github.com/mgieseki/dvisvgm/issues/64) there were issues with libraries or header files for TeX Live I can't handle with :)

I've found another issue while testing TeX Live 2021. The following file fails to render, however in TeX Live 2020 it's rendered fine:

```
\documentclass[11pt]{article}
\usepackage[paperwidth=180in,paperheight=180in]{geometry}
\batchmode
\usepackage[utf8]{inputenc}
\usepackage{amsmath}
\usepackage{amssymb}
\usepackage{stmaryrd}
\newcommand{\R}{\mathbb{R}}

\usepackage[verbose]{newunicodechar}

\newunicodechar{¬}{\ensuremath{\neg}}
\newunicodechar{Γ}{\ensuremath{\Gamma}}
\newunicodechar{γ}{\ensuremath{\gamma}}
\newunicodechar{λ}{\ensuremath{\lambda}}
\newunicodechar{φ}{\ensuremath{\varphi}}
\newunicodechar{ψ}{\ensuremath{\psi}}
\newunicodechar{ϕ}{\ensuremath{\varphi}}
\newunicodechar{ᵢ}{\ensuremath{{}_{i}}}
\newunicodechar{₀}{\ensuremath{{}_{0}}}
\newunicodechar{₁}{\ensuremath{{}_{1}}}
\newunicodechar{₂}{\ensuremath{{}_{2}}}
\newunicodechar{₃}{\ensuremath{{}_{3}}}
\newunicodechar{₄}{\ensuremath{{}_{4}}}
\newunicodechar{₅}{\ensuremath{{}_{5}}}
\newunicodechar{₆}{\ensuremath{{}_{6}}}
\newunicodechar{₇}{\ensuremath{{}_{7}}}
\newunicodechar{₈}{\ensuremath{{}_{8}}}
\newunicodechar{₉}{\ensuremath{{}_{9}}}
\newunicodechar{ₙ}{\ensuremath{{}_{n}}}
\newunicodechar{ℓ}{\ensuremath{\ell}}
\newunicodechar{→}{\ensuremath{\rightarrow}}
\newunicodechar{⇒}{\ensuremath{\supset}}
\newunicodechar{⇔}{\ensuremath{\Leftrightarrow}}
\newunicodechar{∅}{\ensuremath{\emptyset}}
\newunicodechar{∈}{\ensuremath{\in}}
\newunicodechar{∘}{\ensuremath{\circ}}
\newunicodechar{∙}{\ensuremath{\bullet}}
\newunicodechar{∧}{\ensuremath{\wedge}}
\newunicodechar{∨}{\ensuremath{\vee}}
\newunicodechar{∼}{\ensuremath{\sim}}
\newunicodechar{≠}{\ensuremath{\neq}}
\newunicodechar{≡}{\ensuremath{\equiv}}
\newunicodechar{⊃}{\ensuremath{\supset}}
\newunicodechar{⊕}{\ensuremath{\oplus}}
\newunicodechar{⊖}{\ensuremath{\ominus}}
\newunicodechar{⊢}{\ensuremath{\vdash}}
\newunicodechar{⊤}{\ensuremath{\top}}
\newunicodechar{⊥}{\ensuremath{\bot}}
\newunicodechar{⊻}{\ensuremath{\veebar}}
\newunicodechar{⟝}{\ensuremath{\vdash}}
\newunicodechar{⬓}{\ensuremath{\square}}
\newunicodechar{Σ}{\ensuremath{\sum}}
\newunicodechar{Π}{\ensuremath{\prod}}
\newunicodechar{ⱼ}{\ensuremath{{}_{j}}}

\usepackage{pgf-umlsd}

\pagestyle{empty}

\setlength{\topskip}{0pt}
\setlength{\parindent}{0pt}
\setlength{\abovedisplayskip}{0pt}
\setlength{\belowdisplayskip}{0pt}

\begin{document}
\begin{minipage}{0.1in}
\strut\special{dvisvgm:raw&lt;!--start {?x} {?y} --&gt;}\begin{sequencediagram}
\newthread[red]{M}{Interface}{}
\newinst[1]{DB}{DB}{}
\newinst[1]{Auth}{Auth}{}
\newinst[1]{m_list}{ESB Listener}{}
\newthread[green]{ESB}{ESB}{}
%
\begin{sdblock}{Add user, sync DB error}{}
\begin{call}{M}{PUT /user}{Auth}{user id}
\mess{Auth}{user.add}{ESB}
\end{call}
\postlevel
\begin{call}{M}{\shortstack{INSERT user}}{DB}{DB Error}
\end{call}
\prelevel\prelevel
\begin{messcall}{ESB}{user.add}{m_list}
   \setthreadbias{east}
   \begin{call}{m_list}{SELECT user}{DB}{Not found}
   \end{call}
   \begin{call}{m_list}{SELECT user}{DB}{Not found}
   \end{call}
   \begin{call}{m_list}{SELECT user}{DB}{Not found}
   \end{call}
   \begin{call}{m_list}{SELECT user}{DB}{Not found}
   \end{call}
   \begin{call}{m_list}{SELECT user}{DB}{Not found}
   \end{call}
   \begin{call}{m_list}{SELECT user}{DB}{Not found}
   \end{call}
   \begin{call}{m_list}{SELECT user}{DB}{Not found}
   \end{call}
   \begin{call}{m_list}{SELECT user}{DB}{Not found}
   \end{call}
   \postlevel
   \begin{call}{m_list}{\shortstack{After 5 min:\\ INSERT user, account}}{DB}{}
   \end{call}
   \setthreadbias{center}
\end{messcall}
\end{sdblock}
\end{sequencediagram}

\end{minipage}
\end{document}
```

```sh
u_tex@tau:~/textest.s2cms.ru/www/tmp$ /var/www/u_tex/tl-2021/bin/x86_64-linux/latex 1
This is pdfTeX, Version 3.141592653-2.6-1.40.23 (TeX Live 2021) (preloaded format=latex)
 restricted \write18 enabled.
entering extended mode
(./1.tex
LaTeX2e <2021-06-01> patch level 1
L3 programming layer <2021-07-12>
(/var/www/u_tex/tl-2021/texmf-dist/tex/latex/base/article.cls
Document Class: article 2021/02/12 v1.4n Standard LaTeX document class
(/var/www/u_tex/tl-2021/texmf-dist/tex/latex/base/size11.clo))
(/var/www/u_tex/tl-2021/texmf-dist/tex/latex/geometry/geometry.sty
(/var/www/u_tex/tl-2021/texmf-dist/tex/latex/graphics/keyval.sty)
(/var/www/u_tex/tl-2021/texmf-dist/tex/generic/iftex/ifvtex.sty
(/var/www/u_tex/tl-2021/texmf-dist/tex/generic/iftex/iftex.sty)))
u_tex@tau:~/textest.s2cms.ru/www/tmp$
u_tex@tau:~/textest.s2cms.ru/www/tmp$
u_tex@tau:~/textest.s2cms.ru/www/tmp$
u_tex@tau:~/textest.s2cms.ru/www/tmp$ /var/www/u_tex/tl-2021/bin/x86_64-linux/dvisvgm 1.dvi
pre-processing DVI file (format version 2)
processing page 1
  PostScript error: undefined in makepattern
  Operand stack:
      --nostringval--  --nostringval--  --nostringval--  false  --nostringval--  0.0  --nostringval--  --nostringval--  --nostringval-
  -  --nostringval--  --nostringval--
 ```

I'm not sure if it's a dvisvg issue. Seems like the `latex` command produces some new dvi code that cannot be processed both in dvisvgm 2.9 and 2.11.1. But the DVI->PS->PDF conversion works fine again.

**Comment by mgieseki** [comment]
The PS error occurs because the PS code inside the DVI file defines several shading patterns which are not supported by dvisvgm. This issue is already fixed too (859da7c30db3ec19a30f06aecaa3f309767b5906) -- at least the error is gone but shading patterns are still unsupported.
There were probably some incompatible changes in the PS driver of PGF that dvisvgm doesn't like. You can work around this by using the dvisvgm driver instead:
```tex
\documentclass[11pt,dvisvgm]{article}
...
 ```

**Comment by mgieseki** [comment]
I've fixed the color bug so that the example of the initial post converts correctly now.
